utils.py

```diff
@@ -0,0 +1,182 @@
+                new_tags.append(tag)
+            else:
+                new_tags.append(tag.replace("B-", "S-"))
+        elif tag.split("-")[0] == "I":
+            if i + 1 < len(tags) and tags[i + 1].split("-")[0] == "I":
+                new_tags.append(tag)
+            else:
+                new_tags.append(tag.replace("I-", "E-"))
+        else:
+            raise Exception("Invalid IOB format!")
+    return new_tags
+
+
+def load_sentences(path, zeros):
+    sentences = []
+    sentence = []
+    for line in open(path, 'r', encoding='utf-8'):
+        line = re.sub("\d", "0", line.rstrip()) if zeros else line.rstrip()
+
+        if not line and len(sentence) > 0:
+
+            if 'DOCSTART' not in sentence[0][0]:
+                sentences.append(sentence)
+            sentence = []
+
+        else:
+            word = line.split()
+            assert len(word) >= 2
+            sentence.append(word)
+
+    if len(sentence) > 0:
+        if 'DOCSTART' not in sentence[0][0]:
+            sentences.append(sentence)
+    return sentences
+
+
+def iob2(tags):
+    for i, tag in enumerate(tags):
+        if tag == "O":
+            continue
+        split = tag.split("-")
+        if len(split) != 2 or split[0] not in ["I", "B"]:
+            return False
+        if split[0] == "B":
+            continue
+        elif i == 0 or tags[i - 1] == "O":  # conversion IOB1 to IOB2
+            tags[i] = "B" + tag[1:]
+        elif tags[i - 1][1:] == tag[1:]:
+            continue
+        else:  # conversion IOB1 to IOB2
+            tags[i] = "B" + tag[1:]
+    return True
+
+
+def update_tag_scheme(sentences, tag_scheme):
+
+    for i, s in enumerate(sentences):
+        tags = [w[-1] for w in s]
+
+        if not iob2(tags):
+            continue
+
+        elif tag_scheme == 'iobes':
+            new_tags = iob_to_iobes(tags)
+            for word, new_tag in zip(s, new_tags):
+                word[-1] = new_tag
+
+
+def prepare_dataset(sentences, word_to_id, char_to_id, tag_to_id, lower=True):
+
+    def f(x): return x.lower() if lower else x
+    data = []
+    for sentence in sentences:
+        str_words = [w[0] for w in sentence]
+        words = [word_to_id[f(w) if f(w) in word_to_id else '<UNK>']
+                 for w in str_words]
+        # Skip characters that are not in the training set
+        chars = [[char_to_id[c if c in char_to_id else '<UNK>'] for c in w]
+                 for w in str_words]
+        caps = []
+
+        for s in str_words:
+            if s.lower() == s:
+                caps.append(0)
+            elif s.upper() == s:
+                caps.append(1)
+            elif s[0].upper() == s[0]:
+                caps.append(2)
+            else:
+                caps.append(3)
+
+    tags = [tag_to_id[w[-1]] for w in sentence]
+
+    data.append({'str_words': str_words, 'words': words, 'chars': chars, 'caps': caps, 'tags': tags, })
+    return data

```

Comment:
if I understand the code correctly, `data` is always a singleton list. Why do you create the variable at all, why don't you simply return:
```suggestion
    return [{'str_words': str_words, 'words': words, 'chars': chars, 'caps': caps, 'tags': tags, }]

```
But do you really want a singleton list? Isn't there any error?

Reply:
You should have added a comment to this (and all other comments) indicating how you addressed this issue (see II.2)